### PR TITLE
Voice chat support

### PR DESCRIPTION
**Pull request by SNMetamorph**
I think it's time to merge this, so players can test and report about problems/their suggestions

### CODE REVIEW

**Comment by SNMetamorph** [comment]
@a1batross @Velaron @mittorn

File: engine/client/cl_parse.c [hunk 1]
```diff
@@ -1696,7 +1699,30 @@ CL_ParseVoiceData
...
-	// TODO: ???
+	int size, idx, frames;
+	unsigned char received[8192];
+
+	idx = MSG_ReadByte( msg ) + 1;
+
+	frames = MSG_ReadByte( msg );
+
+	size = MSG_ReadShort( msg );
+	size = Q_min( size, 8192 );
```
Comment (a1batross) [inline engine/client/cl_parse.c:1710]: Use `sizeof( received )` here

File: engine/client/cl_parse.c [hunk 2]
```diff
@@ -3127,6 +3154,7 @@ void CL_ParseLegacyServerMessage( sizebuf_t *msg, qboolean normal_message )
 			break;
 		case svc_voicedata:
 			CL_ParseVoiceData( msg );
```
Comment (a1batross) [inline engine/client/cl_parse.c:3156]: Will legacy servers send `svc_voicedata`?

Highly doubt about it.

File: engine/client/cl_view.c [hunk 3]
```diff
@@ -286,9 +286,6 @@ qboolean V_PreRender( void )
 	if( !ref.initialized )
 		return false;
 
-	if( host.status == HOST_NOFOCUS )
```
Comment (a1batross) [inline engine/client/cl_view.c:289]: What's the reason of this change?

File: engine/client/s_main.c [hunk 4]
```diff
@@ -1860,6 +1863,33 @@ void S_SoundInfo_f( void )
...
 }
 
+/*
+=================
+S_VoiceRecordStart_f
+=================
+*/
+void S_VoiceRecordStart_f( void )
+{
+	if( cls.state != ca_active || cls.legacymode )
```
Comment (a1batross) [inline engine/client/s_main.c:1873]: Since there is a check for legacy mode servers, we absolutely don't need any `svc_voicedata/init` parsing for legacy protocol above.

File: engine/client/voice.c [hunk 5]
```diff
@@ -0,0 +1,368 @@
+#include "voice.h"
```
Comment (a1batross) [inline engine/client/voice.c:1]: Add GPL notice for yourself and @Velaron

File: engine/client/voice.c [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+
+wavdata_t *input_file;
+fs_offset_t input_pos;
+
+voice_state_t voice;
+
+CVAR_DEFINE_AUTO( voice_enable, "1", FCVAR_ARCHIVE, "enable voice chat" );
```
Comment (a1batross) [inline engine/client/voice.c:8]: This and other client-specific cvars probably should FCVAR_PRIVILEGED, so server will not change them arbitrarily.

File: engine/client/voice.c [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+	}
+
+	if ( !Voice_IsRecording() )
+		voice.is_recording = VoiceCapture_RecordStart();
+
+	if ( Voice_IsRecording() )
+		Voice_Status( -1, true );
+}
+
+void Voice_AddIncomingData( int ent, byte *data, uint size, uint frames )
```
Comment (a1batross) [inline engine/client/voice.c:309]: Isn't second parameter better be `const`, so you avoid cast later?

File: engine/client/voice.h [hunk 8]
```diff
@@ -0,0 +1,70 @@
+#ifndef VOICE_H
```
Comment (a1batross) [inline engine/client/voice.h:1]: Add GPL notice for you and @Velaron

File: engine/platform/stub/s_stub.c [hunk 9]
```diff
@@ -94,5 +94,20 @@ void SNDDMA_Shutdown( void )
 	}
 }
 
+qboolean VoiceCapture_Init( void )
```
Comment (a1batross) [inline engine/platform/stub/s_stub.c:97]: Add appropriate stubs to `engine/platform/linux/s_alsa.c` as well

File: engine/server/sv_client.c [hunk 10]
```diff
@@ -2561,6 +2561,58 @@ void SV_ParseCvarValue2( sv_client_t *cl, sizebuf_t *msg )
...
+	MSG_ReadBytes( msg, received, size );
+
+	for( i = 0, cur = svs.clients; i < svs.maxclients; i++, cur++ )
+	{
+		if ( cur->state < cs_connected && cl != cur )
+			continue;
+		
+		length = size;
+
+		if ( MSG_GetNumBytesLeft( &cur->datagram ) < length + 6 )
```
Comment (a1batross) [inline engine/server/sv_client.c:2602]: Wtf `length + 6` means here?

File: engine/server/sv_client.c [hunk 11]
```diff
@@ -2561,6 +2561,58 @@ void SV_ParseCvarValue2( sv_client_t *cl, sizebuf_t *msg )
...
+	client = cl - svs.clients;
+
+	if ( size > sizeof( received ) )
+	{
+		Con_DPrintf( "SV_ParseVoiceData: invalid incoming packet.\n" );
+		SV_DropClient( cl, false );
+		return;
+	}
+
+	if ( !Cvar_VariableInteger( "sv_voiceenable" ) )
```
Comment (a1batross) [inline engine/server/sv_client.c:2590]: You can avoid cvar lookup here with just `sv_voiceenable.value`

File: engine/server/sv_client.c (hunk 10 above)
Comment (a1batross) [reply engine/server/sv_client.c:2602]: At least, leave a comment that 6 is a number of bytes for other parts of message. :)

File: engine/server/sv_init.c [hunk 12]
```diff
@@ -386,6 +386,18 @@ void SV_CreateResourceList( void )
...
+/*
+================
+SV_WriteVoiceCodec
+================
+*/
+void SV_WriteVoiceCodec( sizebuf_t *msg )
+{
+	MSG_BeginServerCmd( msg, svc_voiceinit );
+	MSG_WriteString( msg, "opus" );
+	MSG_WriteByte( msg, Cvar_VariableInteger( "sv_voicequality" ));
```
Comment (a1batross) [inline engine/server/sv_init.c:398]: Same, you can just do `sv_voicequality.value` instead.

File: engine/wscript [hunk 13]
```diff
@@ -165,8 +165,10 @@ def build(bld):
 			'client/*.c',
 			'client/vgui/*.c',
 			'client/avi/*.c'])
+	
+		libs.append('opus')
 
-	includes = ['common', 'server', 'client', 'client/vgui', 'tests', '.', '../public', '../common', '../filesystem', '../pm_shared' ]
+	includes = ['common', 'server', 'client', 'client/vgui', 'tests', '.', '../public', '../common', '../filesystem', '../pm_shared', '../3rdparty/opus/opus/include' ]
```
Comment (a1batross) [inline engine/wscript:171]: Make opus include path is part of it's uselib, so it will be added automatically by waf. See `sdl2.py` for manual usage of uselib.

File: wscript [hunk 14]
```diff
@@ -54,6 +54,7 @@ class Subproject:
 		return True
 
 SUBDIRS = [
+	Subproject('3rdparty/opus'),
```
Comment (a1batross) [inline wscript:57]: Probably better be `mandatory = True`? For now, this code doesn't seem to have alternative code path for no voice support anyway.

**Review by a1batross — CHANGES_REQUESTED** [review]

File: engine/client/cl_view.c (hunk 3 above)
Comment (SNMetamorph) [reply engine/client/cl_view.c:289]: Guess it came from Velaron's fork

Comment (Velaron) [reply engine/client/cl_view.c:289]: IIRC, it was causing sound issues when the game window is inactive (not minimized) as I was testing the voice chat basically with myself on the same machine

File: engine/client/cl_parse.c (hunk 1 above)
Comment (SNMetamorph) [reply engine/client/cl_parse.c:1710]: Maybe allocating such a big buffer on stack is a bad idea? I can make it static

File: engine/client/cl_parse.c (hunk 2 above)
Comment (SNMetamorph) [reply engine/client/cl_parse.c:3156]: Yes, I think too, will remove this

File: engine/wscript (hunk 13 above)
Comment (SNMetamorph) [reply engine/wscript]: I tried:
```
def configure(conf):
	if conf.check_pkg('opus', 'OPUS', OPUS_CHECK, fatal = False):
		conf.env.HAVE_OPUS = True
		return

	if not conf.path.find_dir('opus') or not conf.path.find_dir('opus/src'):
		conf.fatal('Can\'t find opus submodule. Run `git submodule update --init --recursive`.')
		return

	conf.env.INCLUDES_OPUS = [
		os.path.abspath('opus/include'),
	]
```
But this doesn't work, what I do wrong?

Comment (a1batross) [reply engine/wscript:171]: Ah, I forgot you're compiling it yourself. Sure, you can do this way.

But you also have export_includes set up already. It should work just right away, I even tested building, and it works without changing include paths for engine.

File: engine/client/cl_parse.c (hunk 1 above)
Comment (a1batross) [reply engine/client/cl_parse.c]: Not at all?

File: engine/client/cl_view.c (hunk 3 above)
Comment (a1batross) [reply engine/client/cl_view.c:289]: @Velaron I don't remember the actual desired behavior in this case.

I think for the sake of debugging, it's better to limit fps and mute only in minimized state. But then NOFOCUS state make no sense.

Comment (SNMetamorph) [reply engine/client/cl_view.c:289]: What do with this? Keep or remove?

File: engine/client/cl_parse.c (hunk 1 above)
Comment (SNMetamorph) [reply engine/client/cl_parse.c]: I made it as static, keep as it is?

Comment (a1batross) [reply engine/client/cl_parse.c]: Use sizeof. It's fine to have 8k buffer on stack, I think we have somewhere even more than that.

File: engine/client/cl_view.c (hunk 3 above)
Comment (a1batross) [reply engine/client/cl_view.c:289]: It's ok.

File: engine/client/cl_parse.c (hunk 1 above)
Comment (SNMetamorph) [reply engine/client/cl_parse.c]: Ok fixed

**Review by a1batross — APPROVED** [review]
